### DISCUSSION TITLE
METplus 6 Rounding Differences

### DISCUSSION THREAD

**Discussion by lindsayblank** [dtcenter/METplus#2831] [Statistical Computation] [answered]
Hello METplus folks,

I am working on upgrading GSL verification workflows from METplus 5/MET 11 to METplus 6/MET 12. So far, 99.9% of the data is identical between the METplus 5 and METplus 6 runs. However, I have found some cases of very small differences in values that look like they are due to rounding. The workflows are run on the same machine (Hera) and on the same data. Unfortunately, I did not record whether or not I ran with METplus 6 beta-6 or METplus 6 rc1, as I used both in testing. I started with beta-6 and moved to rc-1 when it became available. 

These differences are so small, that we are not necessarily concerned about their impact on our verification statistics, but we thought you would want to know.

Here are the differences for context. All of these numbers are from the SL1L2 line type and were found among different forecast models.
METplus 6: 275619382.11474 METplus 5: 275619382.11473
METplus 6: 479045806.12288 METplus 5: 479045806.12289
METplus 6: 105894411.63212 METplus 5: 105894411.63213
METplus 6: 144474736.71988 METplus 5: 144474736.71989
METplus 6: 83958385.64757 METplus 5: 83958385.64758
METplus 6: 144474736.71988 METplus 5: 144474736.71988
METplus 6: 144474736.71988 METplus 5: 144474736.71989
METplus 6: 561697016.01394 METplus 5: 561697016.01395
METplus 6: 144474736.71988 METplus 5: 144474736.71989
METplus 6: 418589968.07748 METplus 5: 418589968.07749

**Comment by lindsayblank** [comment]
Update, I found another rounding difference and now I know exactly which version: METplus 6.0.0-rc1/MET 12.0.0-rc1.

**Reply by JohnHalleyGotway** [reply, marked as answer]
@lindsayblank, I certainly agree that these very small numeric differences out in the 5th decimal place are not a reason for concern.

Many years ago we ran into issues with checking for equality of floating point numbers and switched to using these [is_eq(...)](https://github.com/dtcenter/MET/blob/c7e4345c384caad0c68c37b80843903897a87697/src/basic/vx_math/is_bad_data.h#L54) functions to do so. They check for the differences of the numbers being within some sort of tolerance. By default, we use a "tight tolerance" of 10E-10 but in other contexts, we use a "loose tolerance" of 10E-5.

I'm not that surprised or concerned about differences in the 5th decimal place of large partial sum values. Since they are sums of squares, very small differences in the raw fcst/obs input values get amplified when they're squared. I do understand that the outputs are both generated on Hera, but there might still be differences in the hardware and OS-specifics of the individual nodes.

 We do run nightly regression tests for MET and are currently still using R to check for differences (METplus uses Python, and we'll eventually switch to that). Here's the [line of code](https://github.com/dtcenter/MET/blob/c7e4345c384caad0c68c37b80843903897a87697/internal/test_unit/R_test/test_util.R#L421) that check for numeric diffs:
```
listDiff = signif(dfV1[[strCol]], intSigFig) - signif(dfV2[[strCol]], intSigFig);
```
It calls R's `signif` function to define the number of significant figures when checking for diffs.

In summary, I'm glad you're paying attention to the output. I'm not concerned about this scale of diffs in partial sums, since they are at the scale of not significant floating point diffs. And if you'd like, I could point you to the Python-based differencing logic METplus uses to flag real differences.

**Reply by lindsayblank** [reply]
Great! I appreciate the explanation. Since no one is concerned, I'll mark this complete.
